anvill/include/anvill/Lifters/Context.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021 Trail of Bits, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <optional>
+
+namespace llvm {
+class Constant;
+class GlobalValue;
+class Module;
+}  // namespace llvm
+namespace remill {
+class Arch;
+}  // namespace remill
+namespace anvill {
+
+class ContextImpl;
+class FunctionLifter;
+class LifterOptions;
+class MemoryProvider;
+class TypeProvider;
+class ValueLifter;
+
+// Lifting context for ANVILL. The lifting context keeps track of the options
+// used for lifting, the module into which lifted objects are placed, and
+// a the mapping between lifted objects and their original addresses in the
+// binary.
+class Context {
+ public:
+  ~Context(void);
```

Comment:
Does this need to be declared explicitly?

Reply:
Insofar as I explicitly define the destructor, yes it does. To the extent that my defined destructor is empty and apparently useless, then no, it does not need to be declared or defined. This is a habit I have developed when using the pimpl pattern. I have this habit for two reasons:

 1. If we switched to a `unique_ptr` down the road in place of a `shared_ptr` then it'd be required to maintain our ability to forward declare but not define the implementation class. I often waffle on shared vs. unique, so this is the stronger of my two reasons.
 2. If we want non-trivial destruction in the future, then we can support it without ABI changes. This is a weaker point, because in practice we wouldn't put logic in `~Context`, but instead in `~ContextImpl`.

Reply:
Well, I'm not a huge fan of the pImpl pattern altogether due to adding an unnecessary layer of encapsulation / indirection between the public header and the implementation of the API. More often than not users of our code will end up reading the implementation due to an exhaustive documentation not being available and the pImpl pattern complicates this. That's just my opinion though, so maybe someone else can chime in on this.